### ISSUE TITLE
Out of Memory for Sparse Node Cache Error

### ISSUE THREAD

**Issue by dbowmans46** [openstreetmap/osm2pgsql#508]
I am setting up a PostGIS enabled PostGreSQL database for the entire planet.osm file.  I am using WIndows 2008 R2 64-bit.  When I run the osm2pgsql code attached in the file [load_osm_data.txt](https://github.com/openstreetmap/osm2pgsql/files/40505/load_osm_data.txt), I get an error as follows:

    Allocating memory for sparse node cache
    Out of memory for sparse node cache, reduce --cache size

The complete console input and output can be found in the following file: 
[osm2pgsql_cygwin_console.txt](https://github.com/openstreetmap/osm2pgsql/files/40657/osm2pgsql_cygwin_console.txt)

(Note: I was unable to run the code as a .sh file from the console, so I just copied and pasted the contents into the console.)

As indicated at the top of the cygwin_console.txt, I have over 14000 MB of available memory, but when trying to allocate under 10600 MB, I get the out of memory error.  I am able to get the script to work if I use 1440 MB or less for the cache, but this will take too long.  I have also tried to hard code in 10000 MB for the cache size in case my code was outputting weird data to the osm2pgsql command, but still got the same out of memory error.

Any ideas as to why I cannot allocate more of the available memory to the osm2pgsql.exe process?

**Comment by pnorman** [comment]
What version of osm2pgsql?

Please provide an exact command line, not a script that generates one.

**Comment by dbowmans46** [comment]
Version 0.85.0 64bit.  

The exact command line history of the script ran from the first file above is shown in the second file uploaded above: osm2pgsql_cygwin_console.txt

**Comment by pnorman** [comment]
> The exact command line history of the script ran from the first file above is shown in the second file uploaded above: osm2pgsql_cygwin_console.txt

No, that still has variables in it. Can you run a command by itself

> Version 0.85.0 64bit.

It shouldn't matter for what you're seeing, but the new 0.89.0-dev MSVC builds from appveyor are probably better to use on Windows. I don't run osm2pgsql on Windows, so I don't know the procedure for installing them.

**Comment by dbowmans46** [comment]
I copy and paste the entire code into the command line without any preexisting variables.  Do you mean to run the single osm2pgsql command without first creating the variables at all?  

I will try using the newer software, as well.  Also, thanks for the assistance.

**Comment by pnorman** [comment]
> I copy and paste the entire code into the command line without any preexisting variables. Do you mean to run the single osm2pgsql command without first creating the variables at all?

Yes.

**Comment by dbowmans46** [comment]
Here is the result of just the single command: 
[no_var_attempt.txt](https://github.com/openstreetmap/osm2pgsql/files/40723/no_var_attempt.txt)

Here a picture of the console, in case it helps: 
![console](https://cloud.githubusercontent.com/assets/15948225/11316209/d5e08dc6-8fce-11e5-96e0-3afb4b40c954.png)

Output of /proc/meminfo for snapshot of memory usage:
![mem_usage](https://cloud.githubusercontent.com/assets/15948225/11316225/3ea71f28-8fcf-11e5-9da6-7e8e18cfbc00.png)

**Comment by pnorman** [comment]
Did you try something between 1440 and 5000?

**Comment by dbowmans46** [comment]
Yes, as stated above I had tried 1440 and it ran.  I also tested again with 2000 MB and it ran, but I would like this to run at the maximum memory due to the large amount of time the upload process takes.

**Comment by pnorman** [comment]
Looks like an overcommit issue. On linux you'd do `do sysctl -w vm.overcommit_memory=1` but I'm not sure what the windows equivalent is.

Can you try adding the --cache-strategy dense` flag and using `--cache 10000`? This should confirm what the problem is.

**Comment by dbowmans46** [comment]
Thanks for the suggestion.  This is for a project at work, and unfortunately, the office is closed for Thanksgiving break.  I will be unable to try until after Thanksgiving break, but I will try first thing next Monday when the office opens.  Again, I appreciate the help and will get back to you when I get results.

**Comment by dbowmans46** [comment]
Here is the output when running with the dense cache-strategy:

![error_osm2pgsql_dense_cache_strategy](https://cloud.githubusercontent.com/assets/15948225/11483581/15b7d972-9777-11e5-9848-be1b12bb5834.png)

**Comment by dbowmans46** [comment]
Any ideas on the issue and/or a possible fix?

**Comment by pnorman** [comment]
No. It might be worth trying the 0.89.0-dev builds from Appveyor. Unfortunately, I can't help you run them since I don't run osm2pgsql on windows.

**Comment by martin-s** [comment]
I had same issue on 0.85 (see issues). It was a bug in windows build which hasn't been fixed for this old version. But, there were a lot of changes which also reflect on the windows build, in a positive way (e.g. threads). 
So, Why don't you try a newer and native windows build?

**Comment by dbowmans46** [comment]
Thanks for the info, guys.  I just now had time to get the 0.89 version, and am getting the dependencies sorted out.  I will post results when I get them.

**Comment by lonvia** [comment]
No more info provided. Assuming it has been fixed in the meantime.

**Comment by dbowmans46** [comment]
Sorry for the long interim, it has been awhile since I worked on this project.  I was never able to solve the memory problem with osm2pgsql, and had to go a different route to get the tiles.  

If the new method is important, I used TileMill to create .mbtiles, then converted to .png files via MBUtil. From there, I could setup a folder hierarchy that OpenLayers was able to read and display on our website.
